pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java

```diff
@@ -544,8 +547,18 @@ protected boolean updateOldSegmentMetadataZNRecord(String realtimeTableName, LLC
     String rawTableName = TableNameBuilder.extractRawTableName(realtimeTableName);
     committingSegmentMetadata.setDownloadUrl(
         ControllerConf.constructDownloadUrl(rawTableName, committingSegmentNameStr, _controllerConf.generateVipUrl()));
-    // Pull segment metadata from incoming segment and set it in zk segment metadata
-    SegmentMetadataImpl segmentMetadata = extractSegmentMetadata(rawTableName, committingSegmentNameStr);
+    // Pull segment metadata from committingSegmentDescriptor if present. Otherwise extract it from the incoming segment
+    // and set it in zk segment metadata.
+    // Note: ideally we should use the interface SegmentMetadata here for the class of segmentMetadata. However,
+    // SegmentMetadataImpl has member like ColumnMetadata which is not in the interface. Adding these members to the
+    // interface will introduce circular dependency between Pinot-Core and Pinot-common.
+    SegmentMetadataImpl segmentMetadata;
+    if (committingSegmentDescriptor.getSegmentMetadata() != null &&
+        committingSegmentDescriptor.getSegmentMetadata() instanceof SegmentMetadataImpl) {
+      segmentMetadata = (SegmentMetadataImpl)committingSegmentDescriptor.getSegmentMetadata();
```

Comment:
Do we need the cast here? Maybe we can just define the field in commiting segment descriptor as SegmentMetadataImpl